daemon/c_run.c

```diff
@@ -26,117 +51,159 @@
 //TODO define in container.h?
 #define CLONE_STACK_SIZE 8192
 
-struct c_run {
-	container_t *container;
+typedef struct c_run_session {
+	c_run_t *run;
+	int fd;
+	pid_t active_exec_pid;
 	int console_sock_cmld;
 	int console_sock_container;
 	int pty_master;
-	pid_t exec_loop_pid;
-	pid_t active_exec_pid;
-	pid_t pty_master_read_pid;
-	pid_t pty_master_write_pid;
 	char *pty_slave_name;
 	int pty_slave_fd;
 	int create_pty;
 	char *cmd;
 	ssize_t argc;
 	char **argv;
+} c_run_session_t;
+
+struct c_run {
+	container_t *container;
+	list_t *sessions;
 };
 
 c_run_t *
 c_run_new(container_t *container)
 {
 	c_run_t *run = mem_new0(c_run_t, 1);
 	run->container = container;
-	run->console_sock_cmld = -1;
-	run->console_sock_container = -1;
-	run->pty_master = -1;
-	run->exec_loop_pid = -1;
-	run->active_exec_pid = -1;
-	run->pty_master_read_pid = -1;
-	run->pty_master_write_pid = -1;
-	run->pty_slave_name = NULL;
-	run->pty_slave_fd = -1;
-	run->create_pty = -1;
-	run->cmd = NULL;
-	run->argc = 0;
-
+	run->sessions = NULL;
 	return run;
 }
 
-void
-c_run_free(c_run_t *run)
+c_run_session_t *
+c_run_session_new(c_run_t *run, int create_pty, char *cmd, ssize_t argc, char **argv,
+		  int session_fd)
 {
-	ASSERT(run);
-	mem_free(run);
+	if (argv == NULL) {
+		ERROR("No command was specified to execute.");
+		return NULL;
+	}
+
+	/* Create a socketpair for communication with the console task */
+	TRACE("Setting up sockets");
+	int cfd[2];
+
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, cfd)) {
+		ERROR_ERRNO("Could not create socketpair for communication with console task!");
+		return NULL;
+	}
+
+	c_run_session_t *session = mem_new0(c_run_session_t, 1);
+	session->fd = session_fd;
+	session->run = run;
+	session->pty_master = -1;
+	session->active_exec_pid = -1;
+	session->pty_slave_name = NULL;
+	session->pty_slave_fd = -1;
+
+	session->console_sock_cmld = cfd[0];
+	session->console_sock_container = cfd[1];
+
+	TRACE("Making cmld console socket nonblocking");
+	fd_make_non_blocking(session->console_sock_cmld);
+	fd_make_non_blocking(session->console_sock_container);
+
+	//Add NULL pointer to end of argv
+	session->argv = NULL;
+
+	if (argc > 0 && argv != NULL) {
```

Comment:
existing, redundant check should be removed

Reply:
done